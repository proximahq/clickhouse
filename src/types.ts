import Stream from 'stream';
import type {Pool} from 'undici';
import {supportedJSONFormats, supportedRawFormats} from './constants';

export type JSONDataFormat = typeof supportedJSONFormats[number];
export type RawDataFormat = typeof supportedRawFormats[number];
export type DataFormat = JSONDataFormat | RawDataFormat;

export type JsonValue =
  | string
  | number
  | boolean
  | JsonObject
  | JsonArray
  | null;
export interface JsonArray extends Array<JsonValue> {}
export type JsonObject = {[Key in string]?: JsonValue};

type QueryStringParams = object | any[];
export interface ClickhouseOptions {
  // Options Clickhouse
  host: string;
  port?: number;
  db?: string;
  protocol: string;
  user?: string;
  password?: string;

  size?: number;
  sessionTimeout?: number;
  compression?: boolean;
  // Options for Pool
  connections?: Pool.Options['connections'];
  keepAliveMaxTimeout?: Pool.Options['keepAliveMaxTimeout'];
  headersTimeout?: Pool.Options['headersTimeout'];
  bodyTimeout?: Pool.Options['bodyTimeout'];
}

export interface QueryParams {
  query?: string;
  params?: QueryStringParams;
  queryId?: string;
}

export interface ObjInterface {
  session_timeout?: number;
  output_format_json_quote_64bit_integers?: number;
  enable_http_compression?: number;
  query_id?: string;
  session_id?: string;
  query?: string;
}
export interface BatchTable {
  table: string;
  items: any[];
}

export interface StreamInsertParams {
  table: string;
  items: Stream.Readable;
  format?: DataFormat;
}

export interface StreamInsertResult {
  query_id: string;
}

export interface Client {
  open: () => void;
  close: () => void;
  query: (
    queryString: string,
    params?: any[],
    queryId?: string,
  ) => Promise<any>;
  selectJson: (
    queryString: string,
    params?: any[],
    queryId?: string,
  ) => Promise<any>;
  insertStream: (q: StreamInsertParams) => Promise<StreamInsertResult>;

  insertBatch: (
    q: BatchTable,
    fallback?: (q: BatchTable) => Promise<any>,
  ) => Promise<any>;
  ping: () => Promise<any>;
}
