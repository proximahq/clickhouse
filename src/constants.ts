export const TRAILING_SEMI = /;+$/;
export const JSON_SUFFIX = 'FORMAT JSON';
export const JSON_EACH_SUFFIX = 'FORMAT JSONEachRow';

export const OK = 'Ok.';

export const defaultOpts = {
  host: 'localhost',
  port: 8123,
  db: 'default',
  protocol: 'http',
  user: '',
  password: '',
  connections: 1024,
  keepAliveMaxTimeout: 128,
  headersTimeout: 0,
  bodyTimeout: 0,
  size: 128,
};

export const streamableJSONFormats = [
  'JSONEachRow',
  'JSONStringsEachRow',
  'JSONCompactEachRow',
  'JSONCompactStringsEachRow',
  'JSONCompactEachRowWithNames',
  'JSONCompactEachRowWithNamesAndTypes',
  'JSONCompactStringsEachRowWithNames',
  'JSONCompactStringsEachRowWithNamesAndTypes',
] as const;

export const singleDocumentJSONFormats = [
  'JSON',
  'JSONStrings',
  'JSONCompact',
  'JSONCompactStrings',
  'JSONColumnsWithMetadata',
  'JSONObjectEachRow',
] as const;

export const supportedJSONFormats = [
  ...singleDocumentJSONFormats,
  ...streamableJSONFormats,
] as const;

export const supportedRawFormats = [
  'CSV',
  'CSVWithNames',
  'CSVWithNamesAndTypes',
  'TabSeparated',
  'TabSeparatedRaw',
  'TabSeparatedWithNames',
  'TabSeparatedWithNamesAndTypes',
  'CustomSeparated',
  'CustomSeparatedWithNames',
  'CustomSeparatedWithNamesAndTypes',
] as const;
