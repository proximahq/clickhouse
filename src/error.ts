import type {Dispatcher} from 'undici';
import dbg from 'debug';

const ERRORED = new RegExp('Code: ([0-9]{2}), .*Exception:');

const log = dbg('proxima:clickhouse-driver:error');

interface ErrorProps {
  statusCode: Dispatcher.ResponseData['statusCode'];
  txt: string;
}

export async function getErrorObj(opts: ErrorProps): Promise<Error> {
  const {statusCode, txt} = opts;
  log('getErrorObj');
  let data = '';

  try {
    log('decode stream');
    data = txt;
  } catch (error) {
    //   @ts-ignore
    data = error?.bufferedData ?? '';
  }

  const err = new Error(`Clickhouse error`);
  //   @ts-ignore
  err.body = data;
  //   @ts-ignore
  err.statusCode = statusCode;

  if (data) {
    const m = data.match(ERRORED);
    if (m) {
      if (m[1] && isNaN(parseInt(m[1])) === false) {
        //   @ts-ignore
        err.code = parseInt(m[1]);
      }

      if (m[2]) {
        //   @ts-ignore
        err.message = m[2];
      }
    }
  }
  //   @ts-ignore
  return err;
}
