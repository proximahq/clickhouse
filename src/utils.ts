import {TRAILING_SEMI} from './constants';
import dbg from 'debug';
import {nanoid} from 'nanoid';
import type {JsonObject} from './types';
const log = dbg('proxima:clickhouse-driver:utils');

export const isEmptyObj = (obj: {}) =>
  Object.keys(obj).length === 0 && obj.constructor === Object;

export const isEmptyArr = (arr: any) => Array.isArray(arr) && arr.length === 0;

export const isNil = (v: any) =>
  v === undefined || v === null || isEmptyObj(v) || isEmptyArr(v) || v === '';

export const cleanupObj = (obj: any = {}) => {
  log('cleanupObj');
  const s = Object.keys(obj).reduce((acc, k) => {
    if (isNil(obj[k])) {
      return acc;
    }
    return {...acc, [k]: obj[k]};
  }, {});
  log('obj: %o', obj);
  log('cleanupObj: %o', s);
  return s;
};

export const cleanup = (str: string) => str.replace(TRAILING_SEMI, '');

export const createPathGen = () => {
  const s = {
    session_timeout: 60,
    output_format_json_quote_64bit_integers: 0,
    enable_http_compression: 1,
  };
  return function create(obj: JsonObject) {
    return createPath({...s, ...obj});
  };
};

export const createPath = (obj: any) => {
  const cleaned = cleanupObj(obj);
  const str = new URLSearchParams(cleaned).toString();
  return !str ? '/' : `/?${str}`;
};

export const genIds = () => {
  const maxInt = 2147483647;
  let nextReqId = 0;
  const str = nanoid(16);
  return function next() {
    nextReqId = (nextReqId + 1) & maxInt;
    return `${str}-${nextReqId.toString(36)}`;
  };
};
