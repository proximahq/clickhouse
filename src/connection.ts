import type {Dispatcher, Pool} from 'undici';
import type {URL} from 'url';
import {getErrorObj} from './error';
import {cleanupObj, genIds} from './utils';
import {IncomingHttpHeaders} from 'http';
import dbg from 'debug';
import {OK} from './constants';
import {debug} from 'console';

const log = dbg('proxima:clickhouse-driver:connection');

const factoryId = genIds();

export type Result<R> = {
  statusCode: Dispatcher.ResponseData['statusCode'];
  headers: Dispatcher.ResponseData['headers'];
  data: R;
};

export type DbIfo = {
  db?: string;
  user?: string;
  password?: string;
};

export type DbIfoHeaders = {
  'Accept-Encoding': '*';
  'X-ClickHouse-Database'?: string;
  'X-ClickHouse-User'?: string;
  'X-ClickHouse-Key'?: string;
};

// because undici lazily loads llhttp wasm which bloats the memory
// TODO: hopefully replace with `import` but that causes segfaults
const undici = () => require('undici');

/**
 * Assertion function to make sure that we have a pool
 * @param pool
 */
function assertHasPool<A>(pool: A): asserts pool is NonNullable<A> {
  if (pool === undefined) {
    throw new Error('Connection has not been opened');
  }
}

/**
 * Open an HTTP connection pool
 */
export class Connection {
  private _pool: Pool | undefined;
  private _sessionIds: string[] = [];
  private _isInFallbackMode = false;
  private _hasSessionId: boolean = true;

  private _dbInfoHeaders: Pool | {};

  constructor(dbInfo: DbIfo = {}) {
    this._dbInfoHeaders = cleanupObj({
      'Accept-Encoding': '*',
      'X-ClickHouse-Database': dbInfo?.db,
      'X-ClickHouse-User': dbInfo?.user,
      'X-ClickHouse-Key': dbInfo?.password,
    });

    log('connection constructed', dbInfo);
  }

  getFallbackMode() {
    return this._isInFallbackMode;
  }

  setFallbackMode() {
    this._isInFallbackMode = true;
  }

  removeFallbackMode() {
    this._isInFallbackMode = false;
  }

  getSessionId() {
    if (this._hasSessionId) {
      return this._sessionIds.shift();
    }
    return factoryId();
  }

  returnSessionId(id: string) {
    if (this._hasSessionId) {
      this._sessionIds.push(id);
    }
  }

  isClosed() {
    return this._pool && this._pool.closed;
  }

  /**
   * Initiates a new connection pool
   * @param url
   * @param options
   * @returns
   */
  open(url: string | URL, options?: Pool.Options) {
    if (this._pool) {
      log('connection already open');
      return;
    }

    const {connections = 128} = options || {};

    this._hasSessionId = connections !== null;

    if (this._hasSessionId) {
      this._sessionIds = Array.from({length: connections || 0}).map(() =>
        factoryId(),
      );
    }

    log('connection opening');
    log('pool url', url);

    this._pool = new (undici().Pool)(url, {
      keepAliveMaxTimeout: 600e3,
      bodyTimeout: null,
      headersTimeout: null,
      ...options,
    });

    this._pool?.on('connect', () => {
      this.removeFallbackMode();
      log('connection opened');
    });
    this._pool?.on('connectionError', () => {
      this.setFallbackMode();
      log('connection error');
    });
    this._pool?.on('drain', () => {
      this.setFallbackMode();
      log('connection drained');
    });
    this._pool?.on('disconnect', () => {
      this.removeFallbackMode();
      log('connection closed');
    });
  }

  async getData(resp: Dispatcher.ResponseData['body']) {
    log('getData %o', resp);
    try {
      const results = await resp.json();
      return {...results, status: 'ok', type: 'json'};
    } catch (error) {
      return {status: 'ok', type: 'plain', txt: resp.text};
    }
  }

  /**
   * Perform a request
   * @param method
   * @param endpoint
   * @param headers
   * @param body
   * @returns
   */
  async raw<Results>(
    method: 'POST' | 'GET',
    endpoint: string,
    headers?: Dispatcher.DispatchOptions['headers'],
    body?: Dispatcher.DispatchOptions['body'],
  ) {
    assertHasPool(this._pool);

    const len = body
      ? {'content-length': Buffer.byteLength(body as string)}
      : {};
    const passedHeaders = {
      'Content-Type': 'application/json',
      ...(len as IncomingHttpHeaders),
      ...(this._dbInfoHeaders as IncomingHttpHeaders),
      ...headers,
    };
    log('RAW %s', endpoint);
    log('RAW method %s', method);
    log('RAW body %O', body);
    log('RAW headers %O', passedHeaders);

    const resp = await this._pool.request({
      path: endpoint,
      method: method,
      headers: passedHeaders,
      body: body,
    });

    log('getData');
    const txt = await resp.body.text();
    if (resp.statusCode !== 200) {
      const e = await getErrorObj({
        statusCode: resp.statusCode,
        txt,
      });
      return Promise.reject({error: e});
    }
    if (!txt) {
      return {status: 'ok', type: 'plain', txt: ''};
    }
    if (txt.trim() === OK) {
      return {status: 'ok', type: 'plain', txt: OK};
    }

    try {
      const res = JSON.parse(txt);
      return {...res, status: 'ok', type: 'json'} as Results;
    } catch (_ignore) {
      debug("can't parse response as JSON");
      debug('%o', _ignore);
      debug('txt %s', txt);
      return {status: 'ok', type: 'plain', txt: txt};
    }
  }

  /**
   * Perform a POST request
   * @param endpoint
   * @param body
   * @param headers
   * @returns
   */
  post<Results>(
    endpoint: string,
    body?: Dispatcher.DispatchOptions['body'],
    headers?: Dispatcher.DispatchOptions['headers'],
  ) {
    if (this.getFallbackMode()) {
      return Promise.reject({
        error: new Error('Connection is in fallback mode'),
      });
    }
    log('POST %s', endpoint);
    return this.raw<Results>('POST', endpoint, headers, body);
  }

  /**
   * Perform a GET request
   * @param endpoint
   * @param body
   * @param headers
   * @returns
   */
  get<Results>(path: string, headers?: Dispatcher.DispatchOptions['headers']) {
    log('GET %s', path);
    return this.raw<Results>('GET', path, headers);
  }

  /**
   * Close the connections
   */
  close() {
    if (this._pool) {
      this._pool.close(() => {
        // ignore close errors
      });
    }

    this._pool = undefined;
  }
}
