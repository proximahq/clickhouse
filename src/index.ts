import {format} from 'sqlstring';
import dbg from 'debug';
import {Connection} from './connection';
import {
  ClickhouseOptions,
  StreamInsertParams,
  BatchTable,
  Client,
} from './types';
import {defaultOpts, JSON_SUFFIX, JSON_EACH_SUFFIX} from './constants';
import {
  cleanupObj,
  cleanup,
  createPathGen,
  genIds,
  encodeValues,
} from './utils';

const log = dbg('proxima:clickhouse-driver:main');
const factoryId = genIds();
const createPath = createPathGen();

export type {Client, ClickhouseOptions};
export const clickhouse = (opts: ClickhouseOptions = defaultOpts): Client => {
  const {
    protocol,
    host,
    port,
    db,
    user,
    password,
    connections,
    keepAliveMaxTimeout,
    headersTimeout,
    bodyTimeout,
  } = opts;

  const client = new Connection({db, user, password});

  const ch = {
    open: () => {
      log('opening connection');

      const p = cleanupObj({
        connections,
        keepAliveMaxTimeout,
        headersTimeout,
        bodyTimeout,
      });
      const u = `${protocol}://${host}:${port}`;
      return client.open(u, p);
    },

    close: () => {
      log('closing connection');
      client && client.isClosed() && client.close();
    },

    query: (queryString: string, params: any[] = [], queryId = factoryId()) => {
      if (!queryString) {
        throw new Error('query is required');
      }
      const executableQuery = format(queryString, params);
      const sessionId = client.getSessionId();
      const path = createPath({
        session_id: sessionId,
        query_id: queryId,
      });
      return client.post(path, executableQuery).finally(() => {
        client.returnSessionId(sessionId as string);
      });
    },

    selectJson: <Results>(
      queryString: string,
      params: any[] = [],
      queryId = factoryId(),
    ) => {
      if (!queryString) {
        throw new Error('query is required');
      }
      const q = cleanup(queryString ?? '');
      const executableQuery = `${format(q, params)} ${JSON_SUFFIX};`;
      const sessionId = client.getSessionId();

      const path = createPath({
        session_id: sessionId,
        query_id: queryId,
      });
      return client.post<Results>(path, executableQuery).finally(() => {
        client.returnSessionId(sessionId as string);
      });
    },

    insertStream: (
      q: StreamInsertParams,
      fallback?: (q: StreamInsertParams, err: any) => Promise<any>,
    ) => {
      const {table, items, format} = q;
      const queryId = factoryId();
      if (!table) {
        throw new Error('`table` is required for batch insert');
      }
      if (!items) {
        throw new Error('`items` are required for batch insert');
      }
      if (!format) {
        throw new Error('`format` is required for batch insert');
      }
      const sessionId = client.getSessionId();
      const path = createPath({
        query: `INSERT INTO ${table} FORMAT ${format}`,
        session_id: sessionId,
        query_id: queryId,
      });

      return client
        .stream(path, {}, encodeValues(items, format))
        .catch(err => {
          if (fallback) {
            return fallback(q, err);
          }
          throw err;
        })
        .finally(() => {
          client.returnSessionId(sessionId as string);
        });
    },

    insertBatch: (
      q: BatchTable,
      fallback?: (q: BatchTable, err: any) => Promise<any>,
    ) => {
      const {table, items} = q;
      const queryId = factoryId();
      if (!table) {
        throw new Error('`table` is required for batch insert');
      }
      if (!items) {
        throw new Error('`items` are required for batch insert');
      }

      const sessionId = client.getSessionId();
      const path = createPath({
        query: `INSERT INTO ${table} ${JSON_EACH_SUFFIX}`,
        session_id: sessionId,
        query_id: queryId,
      });

      const executableQuery = JSON.stringify(items);
      return client
        .post(path, executableQuery)
        .catch(err => {
          if (fallback) {
            return fallback(q, err);
          }
          throw err;
        })
        .finally(() => {
          client.returnSessionId(sessionId as string);
        });
    },
    ping: () => {
      return client.get('/ping');
    },
  };
  return ch;
};
