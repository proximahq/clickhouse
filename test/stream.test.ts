import {expect, test, beforeAll, afterAll} from 'vitest';
import {clickhouse} from '../src';
import {dbName} from './utils';
import Stream from 'stream';
import path from 'path';
import fs from 'fs';

const database = dbName();
const config = {
  host: 'localhost',
  protocol: 'http',
  user: 'default',
  port: 8123,
  password: 'password',
  connections: 10,
};

beforeAll(async () => {
  try {
    const ch = clickhouse(config);
    await ch.open();
    await ch.query(`DROP DATABASE IF EXISTS ${database}`);
    await ch.query(`CREATE DATABASE ${database}`);
    await ch.close();
  } catch (error) {
    console.log(error);
  }
});

afterAll(async () => {
  const ch = clickhouse(config);
  await ch.open();
  await ch.query(`DROP DATABASE IF EXISTS ${database}`);
  await ch.close();
});

test('insertStream throws', async () => {
  const client = clickhouse({...config, db: database});
  await client.open();
  await expect(() =>
    // @ts-ignore
    client.insertStream('foo'),
  ).toThrowErrorMatchingInlineSnapshot(
    '"`table` is required for batch insert"',
  );

  await expect(() =>
    // @ts-ignore
    client.insertStream({
      table: 'foo',
    }),
  ).toThrowErrorMatchingInlineSnapshot(
    '"`items` are required for batch insert"',
  );

  await client.close();
});

test('insertStream works for named CSV', async () => {
  const client = clickhouse({...config, db: database});
  await client.open();
  const list = [
    'DROP TABLE IF EXISTS stream_csv_named',
    `CREATE TABLE stream_csv_named
        (
            timestamp DateTime default now(),
            a String,
            b String,
            c Int
        )
        ENGINE = MergeTree()
        ORDER BY timestamp;
        `,
  ];

  for (const query of list) {
    const r = await client.query(query);
    await expect(r).eqls({
      txt: '',
      status: 'ok',
      type: 'plain',
    });
  }

  await client.insertStream({
    table: 'stream_csv_named',
    format: 'CSVWithNames',
    items: fs.createReadStream(
      path.join(__dirname, 'fixtures', 'csv_named.csv'),
    ),
  });

  const res = await client.selectJson('SELECT * FROM stream_csv_named');
  expect(res.data).toMatchSnapshot();
  await client.close();
});

test('insertStream works for plain CSV', async () => {
  const client = clickhouse({...config, db: database});
  await client.open();
  const list = [
    'DROP TABLE IF EXISTS stream_csv_plain',
    `CREATE TABLE stream_csv_plain
        (
            timestamp DateTime default now(),
            a String,
            b String,
            c Int
        )
        ENGINE = MergeTree()
        ORDER BY timestamp;
        `,
  ];

  for (const query of list) {
    const r = await client.query(query);
    await expect(r).eqls({
      txt: '',
      status: 'ok',
      type: 'plain',
    });
  }

  await client.insertStream({
    table: 'stream_csv_plain',
    format: 'CSV',
    items: fs.createReadStream(path.join(__dirname, 'fixtures', 'csv.csv')),
  });

  const res = await client.selectJson(
    'SELECT * FROM stream_csv_plain ORDER BY timestamp',
  );
  expect(res.data).toMatchSnapshot();
  await client.close();
});

test('insertStream works for TSV', async () => {
  const client = clickhouse({...config, db: database});
  await client.open();
  const list = [
    'DROP TABLE IF EXISTS stream_tsv_with_names',
    `CREATE TABLE stream_tsv_with_names
        (
            id Int,
            timestamp Datetime default now(),
            value Int
        )
        ENGINE = MergeTree()
        ORDER BY timestamp;
        `,
  ];

  for (const query of list) {
    const r = await client.query(query);
    await expect(r).eqls({
      txt: '',
      status: 'ok',
      type: 'plain',
    });
  }
  const headers = `id\ttimestamp\tvalue\n`;
  const values = `42\t2013-11-21 20:00:00\t22\n43\t2013-11-21 20:00:01\t55\n`;

  const rawStream = Stream.Readable.from(values, {
    objectMode: false,
  });
  const streamWithNames = Stream.Readable.from(headers + values, {
    objectMode: false,
  });

  await client.insertStream({
    table: 'stream_tsv_with_names',
    format: 'TabSeparatedWithNames',
    items: streamWithNames,
  });

  await client.insertStream({
    table: 'stream_tsv_with_names',
    format: 'TabSeparated',
    items: rawStream,
  });

  const res = await client.selectJson(
    'SELECT * FROM stream_tsv_with_names ORDER BY timestamp',
  );
  expect(res.data).toMatchSnapshot();
  await client.close();
});

test('insertStream works for multiple streams', async () => {
  const client = clickhouse({...config, db: database});
  await client.open();

  const values = `42\t2013-11-21 20:00:00\t22\n43\t2013-11-21 20:00:01\t55\n`;
  const list = [
    'DROP TABLE IF EXISTS stream_multi',
    `CREATE TABLE stream_multi
      (
          id Int,
          timestamp Datetime default now(),
          value Int
      )
      ENGINE = MergeTree()
      ORDER BY timestamp;
      `,
  ];

  for (const query of list) {
    const r = await client.query(query);
    await expect(r).eqls({
      txt: '',
      status: 'ok',
      type: 'plain',
    });
  }

  await Promise.all([
    client.insertStream({
      table: 'stream_multi',
      format: 'TabSeparated',
      items: Stream.Readable.from(values, {
        objectMode: false,
      }),
    }),
    client.insertStream({
      table: 'stream_multi',
      format: 'TabSeparated',
      items: Stream.Readable.from(values, {
        objectMode: false,
      }),
    }),
  ]);

  const res = await client.selectJson(
    'SELECT * FROM stream_multi ORDER BY timestamp',
  );
  expect(res.data).toMatchSnapshot();
  await client.close();
});
