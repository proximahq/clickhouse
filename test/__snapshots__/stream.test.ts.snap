// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`insertStream works for TSV 1`] = `
[
  {
    "id": 42,
    "timestamp": "2013-11-21 20:00:00",
    "value": 22,
  },
  {
    "id": 42,
    "timestamp": "2013-11-21 20:00:00",
    "value": 22,
  },
  {
    "id": 43,
    "timestamp": "2013-11-21 20:00:01",
    "value": 55,
  },
  {
    "id": 43,
    "timestamp": "2013-11-21 20:00:01",
    "value": 55,
  },
]
`;

exports[`insertStream works for multiple streams 1`] = `
[
  {
    "id": 42,
    "timestamp": "2013-11-21 20:00:00",
    "value": 22,
  },
  {
    "id": 42,
    "timestamp": "2013-11-21 20:00:00",
    "value": 22,
  },
  {
    "id": 43,
    "timestamp": "2013-11-21 20:00:01",
    "value": 55,
  },
  {
    "id": 43,
    "timestamp": "2013-11-21 20:00:01",
    "value": 55,
  },
]
`;

exports[`insertStream works for named CSV 1`] = `
[
  {
    "a": "Hi",
    "b": "Hello",
    "c": 3,
    "timestamp": "2013-11-21 20:00:00",
  },
  {
    "a": "Foo",
    "b": "Bar",
    "c": 4,
    "timestamp": "2013-11-21 20:00:01",
  },
]
`;

exports[`insertStream works for plain CSV 1`] = `
[
  {
    "a": "Bar a",
    "b": "Foo B",
    "c": 4,
    "timestamp": "2013-11-11 20:00:01",
  },
  {
    "a": "Name a",
    "b": "Name b",
    "c": 3,
    "timestamp": "2013-11-21 20:00:00",
  },
]
`;
